aten/src/ATen/native/ReduceOps.cpp

```diff
@@ -226,12 +224,116 @@ Tensor cumprod_backward(const Tensor& grad, const Tensor& input, int64_t dim) {
+
+    dF / dx_z1 = sum_{z1 <= j < z2} grad_output[j] * (dy_j / dx_z1)
 
-    so
+    This case has a subtlety though. To compute (dy_j / dx_z1), we cannot use the formula
 
-    grad_output = sum_scan_exclusiv(grad_output * output) / input
+    dy_j / dx_z1 = y_j / x_z1
+
+    as, y_j = x_z1 = 0 for j >= z1. We to compute it with the formula for its derivative, that is:
+
+    dy_j / dx_z1 = prod(x[:z1]) * (grad_output[z1] + sum(grad_output[z1+1:z2] * cumprod(x[z1+1:z2])))
+  */
+
+  if (input.numel() <= 1) {
+    return grad;
+  }
+  dim = at::maybe_wrap_dim(dim, input.dim());
+  const int64_t dim_size = input.size(dim);
+  if (dim_size == 1) {
+    return grad;
+  }
 
+  const auto w = output * grad;
+  const auto is_zero = input == 0;
+  if (!(is_zero.any().item<uint8_t>())) {
+    return reversed_cumsum(w, dim).div(input);
+  }
+
+  // If we are not computing a second order gradient, we can use an
+  // O(n) implementation. The derivative of this implementation is _not_
+  // the second derivative of cumprod. As such, we fallback to a less efficient
+  // O(n^2) implementation when at::GradMode::is_enabled().
+  Tensor grad_input = at::zeros(input.sizes(), grad.options());
+  if (!at::GradMode::is_enabled()) {
+    // n.b. This could probably be implemented much faster with a kernel
+
+    // From here on we need to use some mask gymnastics to
+    // account for the tensorial dimensions
+    // We do a cumsum of the zeros along the dimension.
+    // For a vector is_zero = [False, True, Fals, True, False]
+    // we would have cumsum = [0, 1, 1, 2, 2]
+    // As such we have (in python code for simplicity)
+    // The mask for the range [0, z1):
+    // cumsum == 0
+    // The indices of the first zero z1 and zeros when
+    // there is no first zero:
+    // indices = (cumsum == 1).max(dim, keepdim=True).indices
+    // The mask for the first zero:
+    // zeros_like(indices).scatter_(dim, indices, 1.) & cumsum == 1
+    // Note that the logic_and with cumsum == 1 accounts
+    // for the case when there is no first zero
+    const auto cumsum = is_zero.cumsum(dim);
+
+    // case k < z1
+    // select everything before the first zero [0, z1)
+    auto mask = cumsum == 0;
+    // equiv to grad_input[mask] = deriv[grad]
+    grad_input.masked_scatter_(mask,
+        reversed_cumsum(w.masked_fill(~mask, 0.), dim).div_(input).masked_select(mask));
+    // select everything from the first zero to the second zero [z1, z2)
+    mask = cumsum == 1;
+
+    // case k = z1
+    // We start by select the first zero [z1]
+    // We locate the indices of the first zero using the max function
+    // We then go from the indices to a mask using scatter_
+    // When there is no zero in the slice, max will return the index 0.
+    // To account for this, we need to apply the mask which is true in the range [z1, z2)
+    const auto first_zero_index = std::get<1>(mask.max(dim, /*keepdim*/ true));
+    const auto first_zero_mask = at::zeros_like(mask)
```

Comment:
`const auto first_zero_mask = at::zeros_like(mask).index_fill_(dim, first_zero_index, 1)`

Reply:
You mean replacing `scatter_` by `index_fill_`, right? The `& mask` should stay.

Reply:
Wait, I think `index_fill_` works here with the output of `max`. I think it is right as it is.